src/openkit-shared/Core/Communication/BeaconSendingCaptureOnState.cs

```diff
@@ -63,28 +67,61 @@ protected override void DoExecute(IBeaconSendingContext context)
             HandleStatusResponse(context, statusResponse);
         }
 
+        private void SendNewSessionRequests(IBeaconSendingContext context)
+        {
+            var newSessions = context.NewSessions;
+
+            foreach(var newSession in newSessions)
+            {
+                if (!newSession.CanSendNewSessionRequest)
+                {
+                    // already exceeded the maximum number of session requests, disable any further data collecting
+                    var currentConfiguration = newSession.BeaconConfiguration;
+                    var newConfiguration = new BeaconConfiguration(0,
+                        currentConfiguration.DataCollectionLevel, currentConfiguration.CrashReportingLevel);
+                    newSession.UpdateBeaconConfiguration(newConfiguration);
+                    continue;
+                }
+
+                var response = context.GetHTTPClient().SendNewSessionRequest();
+                if (response != null)
+                {
+                    var currentConfiguration = newSession.BeaconConfiguration;
+                    var newConfiguration = new BeaconConfiguration(response.Multiplicity,
+                        currentConfiguration.DataCollectionLevel, currentConfiguration.CrashReportingLevel);
+                    newSession.UpdateBeaconConfiguration(newConfiguration);
+                }
+                else
+                {
+                    // did not retrieve any response from server, maybe the cluster is down?
+                    newSession.DecreaseNumNewSessionRequests();
+                }
+            }
+        }
+
         private void SendFinishedSessions(IBeaconSendingContext context)
         {
             // check if there's finished Sessions to be sent -> immediately send beacon(s) of finished Sessions
-            var finishedSession = context.GetNextFinishedSession();
-            while (finishedSession != null)
+            var finishedSessions = context.FinishedAndConfiguredSessions;
+
+            foreach (var finishedSession in finishedSessions)

```

Comment:
when using 
var session in finishedSessions
the following code it is easier to read